saleor/shipping/models.py

```diff
@@ -123,17 +124,28 @@ def applicable_shipping_methods_for_instance(
         price: Money,
         country_code=None,
     ):
+        from ..order.utils import check_shipping_method_for_zip_code
+
         if not instance.is_shipping_required():
             return None
         if not instance.shipping_address:
             return None
 
-        return self.applicable_shipping_methods(
+        applicable_methods = self.applicable_shipping_methods(
             price=price,
             channel_id=channel_id,
             weight=instance.get_total_weight(),
             country_code=country_code or instance.shipping_address.country.code,
         )
+        excluded_methods_by_zip_code = []
+        for method in applicable_methods:
+            if check_shipping_method_for_zip_code(
+                instance.shipping_address.postal_code, method
+            ):
```

Comment:
Did you check the performance of this piece of code? I see a few issues here:
- `instance.shipping_address` called in a loop - can result in duplicated database queries
- `check_shipping_method_for_zip_code` calls `method.zip_code_rules.all()` which again is called within a loop so again, potentially duplicated queries.

All data should be prefetched earlier (possibly in GraphQL resolvers) and passed down the function calls so we don't have any database calls in utility functions. I just want to point this out, since there are many more optimization issues on `master` now and we're working on a PR that will fix this. CC @d-wysocki.

Reply:
@maarcingebala `instance.shipping_address` is called in a loop but we loop over `applicable_methods` instead of instances so fortunately that doesn't call extra queries. As for `method.zip_code_rules.all()` that should be better now, I've added prefetching (for now at `applicable_shipping_methods_for_instance` level) and ran the tests - queries count dropped.

Reply:
Ok thanks, performance is super important so we need to make sure that we pay attention to it in all our changes.